Add referencing, namespaces, and extending into base spec

smizell opened the pull request:
Adds in referencing (with ID and Ref), namespaces, and extending elements. 

@fosrias @zdne

smizell committed:
Add referencing, namespaces, and extending into base spec

fosrias commented on refract-spec.md:
```diff
@@ -30,14 +30,10 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
-- `attributes` (enum)
+- `attributes` (object)
 
     The `attributes` property defines attributes about the given instance of the element, as specified by the `element` property. It MAY be an array of elements, or it MAY be an object. The keys of the object SHOULD be defined element types with valid values for the corresponding element types.
```
I think you should mention the reserved properties of attributes here with links to their pertinent definitions, id, namspaces and prefix, etc.

smizell replied:
Added.

fosrias commented on refract-spec.md:
```diff
@@ -30,14 +30,10 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
-
 - `content` (enum)
 
     The `content` property defines the content of the instance of the specified element type. The value MAY be any of the JSON primitive types, an element, or an array of elements.
```
I thought we were going to take out the reference to JSON here and just define these primitive types.

smizell replied:
Yep, but I'm sure I left some in here like this one. Fixed.

fosrias commented on refract-spec.md:
```diff
@@ -54,7 +50,7 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
 #### Example
 
-An simple element may look like this, where `foo` is the element type, `id` is an attribute for the `foo` element, and `content` is a string with a value of `bar`.
+An element may look like this, where `foo` is the element type, `id` is an attribute for the `foo` element, and `content` is a string with a value of `bar`.
```
I think you should change this given that id='4' is probably not compliant with and id element that probably should be a string. Not sure, but at the very least, not sure this is a good example anymore.

smizell replied:
Changed.

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+A `Null Element` is a type that simply has `null` as its value. It extends upon the `Refract Element` and defines a specific content of `null`.
+
+#### Properties
+- `element`: null (string, fixed)
```
So, I sort of wonder if we should highlight attributes here as optional just for completeness.

fosrias replied:
Here and elsewhere

smizell replied:
Put it everywhere an attributes object may be?

fosrias replied:
Yes.
- `attributes` (object, optional)

Albeit the optional part is strange w/r to compact form, but can go with optional here.

smizell replied:
Doesn't each element each element inherit the attributes from the base element?

fosrias replied:
Yes. What is the base element? 

```
{
  "element": "element",
  "attributes": { id, namespaces, prefix },
  "content": ?
}
```

or something?

smizell replied:
That's right.

fosrias replied:
So, then is every other primitive type definable using the base element and extend? That should work right and be the definition of them, basically using an id? Like:

``` json
{ 
  "element": "extend",
  "attributes": { "id": "string" },
  "content": [
    { 
      "element": "element"
    }
  ]
}
```

smizell replied:
In the MSON world, I believe so.

fosrias replied:
Well, we discussed that refract needs to be able to define new types, and that extend was the vehicle for that. That is my recollection.

fosrias replied:
So that I can define element foo and do

``` json
{
  "element": "foo"
}
```

fosrias replied:
Which has different meaning than:

``` json
{
  "element": "ref",
  "content": "#foo"
}
```

smizell replied:
Extend can define new types, but it doesn't always define a new type unless in the right context, correct?

fosrias replied:
So actually, I think the rule is that id forms a reference. If you use it with a ref element it transcludes. If you use it in an element, it defines a type, right?

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+
+### Ref (Element)
+
+The `ref` element can be used to reference elements in remove documents or elements in the local document.
```
sp remote documents

smizell replied:
Fixed

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+
+## Namespacing
+
+Namespaces provide a way to include element definitions and constraints into another document. Namespaces do not include the elements and the data in the elements.
```
... nor the data in the elements.

smizell replied:
Fixed

fosrias replied:
I think you need to discuss the scope of the namespace being the element it is introduced in and all child elements, etc.

smizell replied:
Fixed

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+
+### Namespaces (Element)
+
+The `namespaces` element is a list of namespace elements.
```
Capitalize Namespace elements ?

smizell replied:
Fixed

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+
+### Extend (Element)
+
+The `extend` element provides a way to merge one or more elements into a new element. The `extend` element MUST NOT affect the elements being extended, but MUST define a new instance of an element.
```
... a way to multiply inherit one or more elements to form a new element.

smizell replied:
multiply inherit? Is that correct?

fosrias replied:
How about:

The `extend` element provides a way to multiply inherit from one or more elements to from a new element. The `extend` element MUST NOT modify the elements being extended, but MUST define a new element.

I took out instance. Since this is defining an element type, I don't think instance is right (even though the result is an element). I think an instance refers more to using the type in an element per the structure not pulling in properties, etc. Make sense?

smizell replied:
Maybe the wrong word, but I'm not sure this is a new type either. A type defines constraints for elements (in my mind), while extend merely merges objects.

fosrias replied:
You are right, it defines a new element. Only with the specification of "id" does extend define a referenceable type.

smizell replied:
ID defines a new type in MSON, I presume, but in a normal document, it simple creates a new element of the same type of the given elements. Is that right?

fosrias replied:
Well, id creates a reference. So, if we add an id to an extend element, that creates a new type, right?

smizell replied:
Only in the context in which types are being defined. In a document in which data is represented, a new data element is created. It seems like there is a distinction in there to keep.

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +133,419 @@ This is how it would expressed in the compact version.
+
+The `extend` element provides a way to merge one or more elements into a new element. The `extend` element MUST NOT affect the elements being extended, but MUST define a new instance of an element.
+
+The `extend` element MUST do a deep merge of the elements found in the `content`, but MUST NOT include any IDs in the final element. Each element found in the `content` MUST be of the same element type.
```
Do we need to define a deep merge in this context for specificity?

fosrias commented on refract-spec.md:
```diff
@@ -30,17 +30,19 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
-        - array[Element]
+    Refract reserves the following attributes:
+
+        1. [ID](#id-element)
```
Don't indent. Code block now

smizell replied:
Fixed

fosrias commented on refract-spec.md:
```diff
@@ -54,13 +56,13 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
 {
   "element": "foo",
   "attributes": {
-    "id": 4
```
You did not fix the xml version in similar fashion below here

smizell replied:
Now I did!

fosrias replied:
Actually, this JSON example is wrong. This is correct: https://github.com/refractproject/refract-spec/blob/smizell/referencing-namespacing-extending/refract-spec.md#examples-5

smizell replied:
@fosrias Let's chat about this. This is an example of a Refract element, so shouldn't this be correct?

fosrias replied:
@smizell The problem this example does not represent this JSON example: https://github.com/refractproject/refract-spec/blob/smizell/referencing-namespacing-extending/refract-spec.md#json-representation but the other example is correct.

zdne replied:
Nitpick: "baz" :arrow_right: `baz` ?

smizell replied:
@zdne Fixed

@fosrias Removed XML/JSON representations

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,423 @@ This is how it would expressed in the compact version.
+
+### Property Element (Element)
+
+A `Property Type` allows for values that can be used in `Object Element` elements.
```
MAY or SHOULD instead of can. Should use this throughout these definitions for consistency.

smizell commented:
@fosrias @zdne Added a type element for creating new types. Extend can be used with it to allow for multiple inheritance as well.

fosrias commented on refract-spec.md:
````diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+}
+```
+
+### Array Type (Element)
````
Array Element (Element)

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+
+### Property Element (Element)
+
+A `Property Type` allows for values that SHOULD be used in `Object Element` elements.
```
Should this be a MAY?

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+#### Properties
+- `element`: property (string, fixed)
+- `attributes`
+    - `name` (string, required) - Name of property
```
I could make a case for the rule that if there is only an id, that it defines the name. Otherwise I may have to do id=thing name=thing. Would be nice to be able to do id=thing in the case I want to reference and define the property name. Name alone could preserve name with different id or allow anonymity?

smizell replied:
This is something we should discuss, as we have missed that we introduced a `name` attribute. Should we do it this way?

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+
+### Extend (Element)
+
+The `extend` element provides a way to a way to multiply inherit one or more elements to form a new element. The `extend` element MUST NOT affect the elements being extended, but MUST define a new instance of an element.
```
and MUST define

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+
+The `extend` element provides a way to a way to multiply inherit one or more elements to form a new element. The `extend` element MUST NOT affect the elements being extended, but MUST define a new instance of an element.
+
+The `extend` element MUST do a deep merge of the elements found in the `content`, but MUST NOT include any IDs in the final element. Each element found in the `content` MUST be of the same element type.
```
NOT include any IDs from the content elements in the final element

`content` MUST derive from the same primitive element type.

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+
+The `extend` element MUST do a deep merge of the elements found in the `content`, but MUST NOT include any IDs in the final element. Each element found in the `content` MUST be of the same element type.
+
+The elements are merged from left to right, meaning the latter elements are merged into the former elements.
```
left to right? maybe ... left to right (bottom up in the example),

fosrias commented on refract-spec.md:
```diff
@@ -137,9 +139,481 @@ This is how it would expressed in the compact version.
+
+## Creating Element Types
+
+Element types MAY be referenced or extended to create new element types. If given an ID, these elements MAY be used within the local namespace or referenced from other namespaces if accessible from a resolvable URL. Element types defined within a namespace take priority over all other namespaced types.
```
I think you need to re-enforce how you reference these custom types and the rules for finding them locally or in namespaces. May be another section in the spec that is better.

smizell replied:
Done

smizell commented:
@fosrias Comments addressed.

zdne commented on refract-spec.md:
```diff
@@ -88,7 +90,7 @@ The example above would be represented like this in idiomatic JSON.
 ### Compact Format
 
-In addition to expressing Refract Elements as objects with `element`, `attributes`, and `content` properties, these elements can be expressed as a tuple.
+In addition to expressing Refract Elements as objects with `element`, `attributes`, and `content` properties, these elements SHOULD be expressed as a tuple.
```
SHOULD :arrow_right: MAY ?

smizell replied:
Changed

zdne commented on refract-spec.md:
```diff
@@ -119,7 +121,7 @@ This example will show a JSON object represented as both a normal Refract Elemen
 ##### Refract Element Representation
 
-It can be expressed in the normal Refract element structure.
+It SHOLUD be expressed in the normal Refract element structure.
```
Typo, besides I do not think it make sense here.

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+The `ref` element MAY be used to reference elements in remote documents or elements in the local document.
+
+1. If the content is an absolute URL with a fragment identifier, the fragment identifer references the ID in the corresponding document.
```
Nit: Since you are discussing `content` property I would say this list should be moved under the `content`

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+#### Properties
+
+- `element` ref (string, fixed)
+- `content` (string) - A URL
```
Shouldn't this read: "A URL or fragment"?

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+A prefix is a way to reference elements in prefixed namespaces.
+
+#### Example
```
Missing formal definition of Prefix element. E.g. 

```
### Properties
- element
- content
```

smizell replied:
Fixed. Also did the same for ID element.

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+#### Members
+
+- (string) - A resolvable URL to a namespace
```
I would say this is missing an explanation between `A resolvable URL to a namespace` and `An href is a resolvable URL to a namespace`. I assume those are the same, the later (in the object) is to define a prefix. 

But please elaborate.

smizell replied:
Done

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+- (string) - A resolvable URL to a namespace
+- (object)
+    - `prefix` (string) - A prefix
+    - `href` (string) - An `href` is a resolvable URL to a namespace
```
Nitpick: I do not think you need to mention `href` in the description

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+## Extending Elements
+
+Elements by default do not include any of the attributes or content from the element in which they reference. An element MAY be extended in order to inherit from other elements.
```
I feel this is really missing an example, or better put, the referencing element above needs an example of "The ref element transcludes the content". Maybe then this would make more sense. Can you please elaborate on it?

smizell replied:
Think of a `ref` element kind of like an `img` tag in HTML. With the `img` tag, you do not embed the the image, but provide a reference to it for the purpose of fetching and including in the graphical representation of the document. An `a` tag links to another document, but an `img` tag transcludes it. 

Similarly, with `ref`, you are simply referencing another element. The libraries should treat the referenced element as if it was actually there in the document, hence the idea behind transclusion.

I'll look at getting an example added.

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+### Extend (Element)
+
+The `extend` element provides a way to a way to multiply inherit one or more elements to form a new element. The `extend` element MUST NOT affect the elements being extended, and MUST define a new instance of an element.
```
"MUST NOT affect" as in the original element is not touched OR the inherited properties of original element can not be modified (e.g. requirement attributes etc)?

smizell replied:
The first, that they don't affect the original. Fixed.

zdne commented on refract-spec.md:
```diff
@@ -137,9 +139,490 @@ This is how it would expressed in the compact version.
+
+The `extend` element MUST do a deep merge of the elements found in the `content`, but MUST NOT include any IDs from the content elements in the final element. Each element found in the `content` MUST derive from the same primitive element type. The elements are merged from first to last.
+
+#### Examples
```
I totally do not get anything in these examples :(

zdne replied:
Frankly, I do not see the reason why Extend element is defined in the first place. We can discuss this over the Skype but since it is not obvious from the spec here I have some doubts ..

smizell replied:
The purpose of the extend is purely for declarative and functional reasons. The only thing extend does is take two objects and merge them together. That's it. 

In this PR, elements never inherit from another element. Other elements may be referenced, and elements may provide an ID for the ability to reference, but elements do not inherit properties from one another. 

Extend provides the means by which you may declare that a new element exists by the merging of one or more elements. It does not create a new type (which can be used as the value of `element`), but rather simply a new element.

So there is one place inheritance happens, and then another place where type declaration happens, and that is with the type element. The type element takes one element and provides an ID that may be used as the value of `element`.

Type and Extend may then be used together for multiple inheritance. Multiple elements may be merged together with extend, and then a new type is declared from that extend.

Hope that all makes sense :)

zdne commented:
Long way we have walked! Good job. I am still missing some clarity on referencing versus extending. Extending in particular seems to be confusing to me. For example, what is the difference between:

``` md
{
  "element": "type",
  "attributes": {
    "id": "foo"
  },
  "content": {
    "element": "ref",
    "content": "#bar"
  }
}
```

and

``` md
{
  "element": "type",
  "attributes": {
    "id": "foo"
  },
  "content": {
    "element": "extend",
    "content": [
      {
        "element": "ref",
        "content": "#bar"
      }
    ]
  }
}
```

Is something I didn't really get

zdne commented:
@smizell can you give me an example of an arbitrary element A, element B that inherits from element A and expansion of element B that retains the reference to element A? I think this would help.

smizell commented:
@zdne On your example above, there is essentially no difference. Extend is used when you want to define a new type based on multiple elements rather than just one. Extend can also be used to simply declare a new element, but it alone does not create a new type.

We **could** have not added extend and just defined the type element as being an array. Instead, we abstracted out this functionality into its own element that can be used for expressing a computed element.

smizell commented:
A quick example on defining element types. Basically, a type definition defines constraints for an element, and does not transclude data when referenced as a namespace. In the document it is defined, though, it is still may be referenced and extended. 

## Element A and Element B

Element A here is declared as a new type based on Refract's string primitive.

``` json
{
  "element": "type",
  "attributes": {
    "id": "a"
  },
  "content": {
    "element": "ref",
    "content": "http://refractnamespace.com/#string"
  }
}
```

Type B is a new type based on the Type A declaration.

``` json
{
  "element": "type",
  "attributes": {
    "id": "b"
  },
  "content": {
    "element": "ref",
    "content": "#a"
  }
}
```

## Expansion

I'm a little unsure what an expansion would look like that references back to its original. Wouldn't this be more of an implementation detail for a library parsing this document? What are the reasons to alter the original document by expanding it?

fosrias commented on refract-spec.md:
```diff
@@ -107,34 +89,511 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+- `element` type (string, fixed)
+- `content` (Element)
+
+#### Example
```
I think it my be illuminating to show how the primitives all (at least one or two) inherit from one element contain the attributes properties.

fosrias commented on refract-spec.md:
````diff
@@ -107,34 +89,511 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
   "content": "bar"
 }
 ```
 
````
@smizell I think it would be useful to show the difference of two elements that use extend and ref and look similar but show the merged final extend version and the transcluded version.

This somehow reminds me about the use of `ref` in the attributes that was used to keep a reference to a transcluded element. I think @zdne needs that and we may have removed that capability. Is it still in the spec?

smizell replied:
@fosrias Yes, we decided to take it out. Do you all think it is necessary? If you have actually altered the document to replace the ref element with the element in which it was referencing, is it really transcluding anymore? Isn't it a new element at that point?

zdne commented on refract-spec.md:
```diff
@@ -107,34 +89,511 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+
+Elements by default do not include any of the attributes or content from the element in which they reference. An element MAY be extended in order to inherit from other elements.
+
+### Extend (Element)
```
@smizell this is still missing `Properties` section with explanation of element's properties (mainly for `content`)

zdne commented on refract-spec.md:
```diff
@@ -107,34 +89,511 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+#### Properties
+
+- `element` type (string, fixed)
+- `content` (Element)
```
I am really missing explanation of the `content` property here

zdne commented:
@smizell So I still can't merge this for two reasons: 
1. Missing explanation of extend element properties
2. Missing explanation of type element content property

zdne commented:
@smizell Also I would like to have clarification on how to inherit a basic type. 

This seems artificial: 

```
{
    "element": "ref",
    "content": "http://refractnamespace.com/#string"
}
```

Also what about my comment on using valid URL segments (e.g. one with `#` prefix) versus no constraints on what identifier can be. In other words, if id is of `hello world` value should the reference content be `#hello%20world` or `#hello world` ?

zdne commented on refract-spec.md:
````diff
@@ -107,34 +89,511 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+
+```json
+{
+  "element": "foo",
````
Also this is totally going against your concept of referencing. You are referencing an id here as any other thing this should be `#foo` or full-blown reference element.

pksunkara commented on refract-spec.md:
````diff
@@ -107,34 +89,516 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+
+```json
+{
+  "element": "string",
````
@smizell Is this example wrong? Because the `element` in the `Properties` say that `element` should be `id` and the value is fixed.

smizell replied:
This is an aspect of Refract that I'm wondering if we should remove from the initial spec, because it confuses things. This aspect is, the `attributes` may be expanded to Refract as well, and may include its own elements.

I think the example is correct in concept, but is quite confusing. I'm going to probably vote we remove this and simply define available attributes. If I forget, this is probably something we should bring up after the other conversations.

Thanks for bringing it up.

pksunkara commented on refract-spec.md:
````diff
@@ -107,34 +89,516 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+
+```json
+{
+  "element": "foo",
````
Similarly here, the MSON above says `element` should be `prefix`.

smizell replied:
@pksunkara My recent commit addresses this issue. Feel free to have a look.

pksunkara replied:
@smizell Your recent comment still doesn't address the deprecancy. The `element` described using MSON still says it should be fixed at `prefix` but the example contains `foo` value for `element`.

smizell commented:
@fosrias Please have a look at this PR now. I've added a section on Primitive Types that define the basis for an infinitely recursive structure. This separates the idea of an Object Type and an Object Element. Hope it makes sense, but I think this is pretty close to satisfying a lot of things.

pksunkara commented on refract-spec.md:
```diff
@@ -117,29 +474,307 @@ This example will show a JSON object represented as both a normal Refract Elemen
 {
-  "element": "foo",
-  "attributes": {},
+  "element": "string",
```
This is not fixed yet. The MSON description of `element` says it should be fixed at value `id` but the example contains value `foo`.

smizell commented:
@pksunkara I think I may have addressed your comments.

@zdne I have added a select element.

fosrias commented on refract-spec.md:
```diff
@@ -12,89 +12,279 @@ This document conforms to RFC 2119, which says:
+## Primitive Types
 
-Refract provides a single structure for data, which will be referred to throughout this document as an element. All elements found in a Refract document SHOULD conform to the element structure as defined here.
+Primitive Types are types used by Refract as the basis for all values and structure. These types use both JSON primitive values along with Refract Elements.
```
I think you really need to bolster this a bit to make clear the delineation between what you are calling a 'Type' and an 'Element'.

Effectively these types are used in elements to define structures, correct?

fosrias replied:
In other words, refract uses these "Types" to describe other structures. If that is what you are saying, I think you need to make that crystal clear and have some examples (probably around objects) to make clear the differentiation.

smizell replied:
These types are the basis for Refract, much like a type is the basis for a programming language or structure like JSON. Refract types can be defined as JSON primitives or Refract Elements, which gives these types the ability to be expanded out into infinity.

fosrias replied:
So, this is one fundamental question nagging at me about this spec. Should we couple it to JSON vs. define it as a concept that is more hash based overall and note that for convenience, we use JSON for examples. E.g. in a ruby implementation is this refract:

``` ruby
{
  element: 'foo',
  content: 'bar'
}
```

IMO refact is an abstract structure that in no way is dependent on JSON. I am just wondering if we should clarify this and keep the spec more fundamental. We don't have to address this in this PR, but I think we should think about it. Frankly my inclination is we show a refract element in a few different languages just to make it clear. Thoughts @smizell ?

smizell commented:
I've added and changed everything I'm going to in this PR. Please review for clarity of concepts, and not not necessarily typos and wording. We can fix those later.

My main concern is, is everything covered here needed to move forward with MSON?

@zdne @pksunkara Note that we changed up how properties are done and used a Member Type with a key/value pair. This makes things a lot clearer along with providing the ability to expand and annotate both keys and values separately.

fosrias commented on refract-spec.md:
```diff
@@ -12,89 +12,286 @@ This document conforms to RFC 2119, which says:
+
+Object Types provide a key/value structure for Refract.
+
+When converting an Object Type to a primitive JSON type, and the Object Type is a JSON object, no processing is needed. If the Object Type is an array, the following rules apply:
```
Again, what if I am in ruby and working with hashes. Do we want to use this type of wording?

fosrias commented on refract-spec.md:
```diff
@@ -12,89 +12,286 @@ This document conforms to RFC 2119, which says:
-        - (string)
+A Null Type in Refract can be either the `null` value or a Null Element.
+
+##### Members
```
Why this vs. just using MSON list right after that defines the enum members? Is this to be more human friendly? It is more verbose than it needs to be from an MSON standpoint.

fosrias replied:
Nm. Missed the comment in between the definitions.

fosrias commented on refract-spec.md:
````diff
@@ -12,89 +12,286 @@ This document conforms to RFC 2119, which says:
+```json
+[
+  {
+    "element": "foo",
````
@smizell I fail to see the value add for this definition as noted above: https://github.com/refractproject/refract-spec/pull/5/files#diff-4eef42b68375cd377e0657ef7ea34f6bR113

It convolutes the idea of an element type with a key of an object and I don't see the benefit for the added complexity. It's like which of these elements is not like any of the others? 

I think a much more declarative definition would be:

If the item of the array is an array, then the array must contain a key and a value element.

```
[
  [ ['key', {}, {}, 'foo'], ['value', {}, {}, 'bar'] ],
  [ ['key', {}, {}, 'hello'], ['value', {}, {}, 'world'] ]
]
```

This is far more declarative makes clear this thing is an object type, and I also think in MSON, based on the idea of variable keys, etc, having metadata on a key is going to be useful, particularly with complex objects, potentially. Adding some metadata to ['foo', {}, {}, 'bar'] and that actually means an object member is limiting in this case.

Further, if I ran across:

```
['array', {}, {}, [[['foo', {}, {}, 'bar']]]]
```

is that content an array of foo element or an object.

I think this is a complete stretch and overloads the definition of an element type in a way that is unique only to this case and therefore IMO, not valuable. 

So, back to the question, what is the value add of this vs. the cost of complexity and confusion?

fosrias commented on refract-spec.md:
````diff
@@ -12,89 +12,286 @@ This document conforms to RFC 2119, which says:
-```xml
-<foo id="4">bar</foo>
+```json
+[
````
So, you have this definition here, which is good per my prior comment, but per prior you also define:

`['foo', {}, {}, 'bar']` which only has meaning in one context as a member, but not anywhere else. So, you would need this example, but that as standalone. When you see this its a member if, means mentally looking at this you are always coupled with asking what is this? An element or member. I know, broken record.

fosrias commented on refract-spec.md:
````diff
@@ -117,29 +485,570 @@ This example will show a JSON object represented as both a normal Refract Elemen
+```json
+{
+  "element": "ref",
+  "content": {
````
So, I have been thinking that this is what we have called a namespace element. Frankly, I think in principle it is more of a reference without the idea of transclusion. Almost a `type` and that it has a different connotation as the content of a ref that transcludes its content and a `namespaces` element that 'transcludes' its constraints. Not sure if I have brought this up, but I think we can find a way to collapse and re-use this thing. Then we get some re-use. Just a matter of describing this.

fosrias commented on refract-spec.md:
```diff
@@ -117,29 +485,570 @@ This example will show a JSON object represented as both a normal Refract Elemen
+- `element` namespaces (string, fixed)
+- `content` (array[Namespace])
+
+### Namespace (enum)
```
As I said, I think could be something different than this that can be the definition of a ref content. In the ref context,
if it is a string, it is the current namespace and href is a relative fragment (with or without # depending on how we define it) points to local element. If href is absolute and no prefix, still has to point to local element. If it has prefix and fragment it lives that namespace, but if it is absolute URI it is that element etc. Transclude it. If it is in namespaces it references constraints.

smizell commented:
@fosrias Mark, I added a special Refracted Object. This is the content of the Object Element and the meta and attributes properties.

The reason for this was because there was previously no way to distinguish between a regular object and an element as the values for meta and attributes since they are really both objects. Now there is a clear way to distinguish, so it seems to fit now.

smizell commented:
@fosrias Update: instead of a new definition, I just changed the definitions for meta, attributes, and Object Element. I think this resolves that issue. Let me know what you think.

zdne commented on refract-spec.md:
```diff
@@ -12,35 +12,46 @@ This document conforms to RFC 2119, which says:
     - Members
-        - (object)
-        - array[Element]
+        - (string)
```
I believe this should be `(object)` (a hash) not a string – attributes is a hash or array of elements vs. attributes is a string or array of elements

smizell replied:
You are correct.

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -50,53 +61,38 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
+### Compact Format
 
-The example above would be represented like this in idiomatic JSON.
+In addition to expressing Refract Elements as objects with `element`, `attributes`, and `content` properties, these elements MUST be expressed as a tuple.
```
missing `meta` ?

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -50,53 +61,38 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
-    - Compact Element
-- (enum, optional) - Attributes of the element
+- (string, required) - Name of the element
+- (enum, optional) - Meta attributes of the element instance. See meta attributes above for full Refract representation
```
I believe this should be `required`

smizell replied:
Fixed, along with all dittos on this section

zdne commented on refract-spec.md:
```diff
@@ -50,53 +61,38 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
     - (object)
-    - array[Compact Element]
+    - (array[Compact Element])
+- (enum, optional) - Attributes of the element instance
```
ditto

smizell replied:
Fixed

zdne commented on refract-spec.md:
```diff
@@ -50,53 +61,38 @@ The Refract Element contains three properties, `element`, `attributes`, and `con
+- (enum, optional) - Attributes of the element instance
+    - (object)
+    - (array[Compact Element])
 - (enum) - Element content with any of the following types
```
ditto

smizell replied:
Fixed

zdne commented on refract-spec.md:
````diff
@@ -105,11 +101,193 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+This is how it would represented in the compact version.
+
+```json
+["foo", {}, {}, "bar"]
````
Not so nice consequence is that `["foo", [], [], "bar"]` is equal to this example.

smizell replied:
Isn't that OK? Isn't that what allows elements to be added to meta and attributes?

zdne replied:
It is OK. It may just be confusing to some, maybe then `["foo", null, null, "bar"]` should be a preferred notation to not hint element or hash.

zdne commented on refract-spec.md:
```diff
@@ -105,11 +101,193 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+- `element`: string (string, fixed)
+- `content` (enum)
+    - (string)
+    - (String Element)
```
What is the reason for explicitly allowing string element in string element content ? (applies for other primitives too)

smizell replied:
It allows for anything in refract to be annotated with meta or attributes. If it is too confusing, I vote taking it out. cc @fosrias

zdne replied:
:+1: for taking this out. I think it is not needed right now and we can always add this later. The simpler the initial version is going to be the better.

zdne commented on refract-spec.md:
```diff
@@ -105,11 +101,193 @@ In addition to expressing Refract Elements as objects with `element`, `attribute
+
+- `element`: object (string, fixed)
+- `content` (enum)
+    - (string)
```
I do not get this? Why (string) here?

smizell replied:
Wires got crossed here. Should be object.

smizell replied:
Was changing this at 1AM :(

zdne commented on refract-spec.md:
```diff
@@ -117,26 +295,334 @@ This example will show a JSON object represented as both a normal Refract Elemen
+
+### Example
+
+The element below is a string element with an ID of "foo."
```
"foo".

smizell replied:
This is a big debate in Apiary right now :) The American way is to put periods inside the quotation marks.

zdne commented:
Done with the review.

zdne commented:
Only minor comments. Then it is good to merge

smizell commented:
Changes made.

zdne commented:
Merged
